google-assistant-sdk/googlesamples/assistant/grpc/pushtotalk.py

```diff
@@ -138,6 +137,7 @@ def iter_assist_requests():
                 logging.info('Transcript of user request: "%s".',
                              ' '.join(r.transcript
                                       for r in resp.speech_results))
+                self.conversation_stream.start_playback()
```

Comment:
I think we want to keep doing this in the same grpc thread from where the stream reading is happening (i.e: at the end of the generator).

Reply:
I think the whole problem is that I get data for playback in assist response before the request generator ends, so the playback is not started yet and writing to playback stream gets stuck while waiting to start the playback...

Reply:
@PizlaTheDeveloper yep, but I don't think this will be an issue if we use two different stream as you introduced with https://github.com/googlesamples/assistant-sdk-python/pull/206/commits/d0eb6a36d1d175d47235fa472a1487e1a06b203a

Can you give a try to https://github.com/googlesamples/assistant-sdk-python/pull/188/commits/2c63ebc1072266ecfbf118810bc54ad6e0279796 and let me know if it works for you?

Reply:
Sorry, but it doesn't work :( 
The generator 'iter_assist_requests' will never stop iterating because it iterates while conversation_stream has data on its input, but there will be always some data if the source stream is running. So the recording will never stop.

I think that the events '_stop_recording' and '_start_playback ' should stay in ConversationStream to be able to signalize stop recording and return empty data to read calling what will end the iteration. Stop recording must be signaled when END_OF_UTTERANCE occurs.

I don't know if I described it correctly :)

I think my previous code was correct :)

Thank you for your interest

Reply:
It sounds good, but unfortunately it is not the right solution :(
Firstly I usually test to start assistant while not speaking anything. So the assistant iterates only two responses - 'END_OF_UTTERANCE' and 'dialog_state_out'. But if these to responses comes faster then 'iter_assist_requests' can iterate then iterating responses will end before iter_assist_requests ends, so the 'stop_recording' is not called again.

Reply:
@PizlaTheDeveloper Thanks for digging this out.

Because of https://github.com/grpc/grpc/blob/16316a9fec3a78baf83d96dbdb392e90226bb013/src/python/grpcio/grpc/_channel.py#L682, I was making the assumption that grpc would always exhaust the request generator before ending the iteration of the response, and thought this would always be called:
https://github.com/googlesamples/assistant-sdk-python/commit/187b485afd5c5df8288a135108065d963a201a84#diff-7262ae17d2cea4f6cb6c2b5934aaf939L127

But it looks like it's not the case! and that the iteration on `self.assistant.Assist` responses can finish before `iter_assist_requests` is exhausted. (Maybe @nathanielmanistaatgoogle can shed some light on the expected behavior).

@PizlaTheDeveloper can you give a try at https://github.com/googlesamples/assistant-sdk-python/pull/188/commits/2dfc868ebce1b78f772afe1af174337cc2de4ebc I added explicit locking to make sure we can safely close the recording stream from the main thread.

Reply:
It's working like a charm now !!!

Reply:
can you review/approve #188 ?

Reply:
Without reading this whole conversation, so I may be missing some context: yes, it is **not** the case that exhaustion of the request iterator is guaranteed. If an RPC ends (for whatever reason, successfully or unsuccessfully) before exhaustion of the request iterator, iteration of the request iterator may be stopped. (Or it may be continued! Don't depend on either being the case.)

Reply:
@nathanielmanistaatgoogle Thanks for the confirmation! So we *really* shouldn't perform any cleanup/teardown logic at the end of the iterator :)